Add and use Pipfile

stephendonner opened the pull request:
@davehunt r?

Passing ad-hoc here: https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/stubattribution.adhoc/159/console

davehunt commented on .travis.yml:
```diff
@@ -1,7 +1,7 @@
 python: 2.7
 install:
-  - pip install -r requirements.txt
+  - pip install flake8
```
I think we can change this to use pipenv. Perhaps `pip install pipenv` for the install and `pipenv check` and `pipenv check --style .` for the script?

davehunt commented on Dockerfile:
```diff
@@ -1,8 +1,8 @@
 RUN pip install pipenv
-RUN pipenv install --system -r requirements.txt --skip-lock --deploy
+COPY Pipfile Pipfile.lock /src/
+RUN pipenv install --system --skip-lock --deploy
```
We should remove `--skip-lock` now that we're using these files.

davehunt commented on Pipfile:
```diff
@@ -0,0 +1,25 @@
+pytest-variables = "==1.7.1"
+pytest-xdist = "==1.22.0"
+"querystringsafe_base64" = "==1.0.0"
+selenium = "==3.8.1"
```
My basic understanding of the Pipfile is that we should not pin here unless it's necessary. It means we can regenerate the Pipfile.lock at any time from the latest versions of dependencies, but only commit the updated lock file once we're happy that the dependencies are good.

davehunt replied:
Add the following section to indicate that Python 2.7 is currently a requirement:

```ini
[requires]
python_version = "2.7"
```

davehunt commented on README.md:
````diff
@@ -74,7 +74,7 @@ required for these tests.
 ```
 $ pip install pipenv
-$ pipenv --two install -r requirements.txt
+$ pipenv --two install
````
This should be even easier now. Try simply `pipenv run pytest` and the virtual environment should be created if if doesn't already exist.

stephendonner commented:
@davehunt thanks for the review; I believe I've addressed your feedback - new ad-hoc run here: https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/stubattribution.adhoc/160/console

davehunt commented on Pipfile:
```diff
@@ -0,0 +1,27 @@
+pytest-selenium = "*"
+pytest-variables = "*"
+pytest-xdist = "*"
+"querystringsafe_base64" = "==1.0.0"
```
Strange that this is in quotes. Is it because of the `_` in the name?

stephendonner replied:
I believe so - was done automagically via pipenv install

davehunt commented on README.md:
````diff
@@ -74,7 +74,7 @@ required for these tests.
 ```
 $ pip install pipenv
-$ pipenv --two install -r requirements.txt
+$ pipenv run pytest
````
This runs the tests, so you can remove the following section.

davehunt commented on requirements.txt:
```diff
@@ -1,12 +0,0 @@
-pytest-variables==1.7.1
-pytest-xdist==1.22.0
-querystringsafe_base64==1.0.0
-selenium==3.8.1
```
If we want to keep pyup support, we need to tell it about the Pipfile by creating a .pyup.yml file:

```yml
requirements:
  - Pipfile
  - Pipfile.lock
```

stephendonner replied:
I've got that in a followup PR, but will fold that in here, as it makes sense; thanks!

stephendonner replied:
Added in https://github.com/mozilla/stubattribution-tests/pull/106/commits/efadb7ef26c7163d452ff8bc1b30dc72834e6daa#diff-50f285e3023eb0b37f3dc8671b7dace9

davehunt commented on Pipfile:
```diff
@@ -0,0 +1,27 @@
+
+"flake8" = "*"
+"flake8-isort" = "*"
+funcsigs = "*"
```
This shouldn't be needed anymore as it's a pytest dependency and pipenv should pick it up. Can you try removing it? This list should be the direct dependencies only, and Pipfile.lock will be the complete list.

stephendonner commented:
Hopefully addressed all review feedback; latest ad-hoc build (passing) is here: https://qa-preprod-master.fxtest.jenkins.stage.mozaws.net/job/stubattribution.adhoc/162/console

davehunt commented on README.md:
````diff
@@ -74,12 +74,6 @@ required for these tests.
 
 ```
 $ pip install pipenv
````
You don't need this line.. `pipenv run` wil create the virtual environment and install the dependencies listed in the `Pipfile.lock` for you.
